Filter nexus delayed hits in detsim

gondiaz opened the pull request:
`detsim` contains a parameter `max_time` such that all hits ocurring afterwards this threshold are not included in the waveform simulation. For example, a nucleus can decay to its ground state via two gamma deexcitations, being the decay halt-life of the latter much larger than the initial. 
In such a case, if the second decay is within the limit set by `max_time`, `detsim` will be able to process both signals and split them in two different events if the second decay occurs at time larger than the `buffer_length` of the waveform.
But if the second decay time is larger than `max_time`, `detsim` is not filtering these hits and an error is raised in the waveform creation.

This PR solves this issue by adding a function that filters the hits at times larger than `max_time`. A warning is raised for these events for the user to be aware that the simulation might not provide the expected output.
This PR also fixes the fact that S1 has to be computed using the hits at the active and buffer volumes, not only in the active as was previously done.

gondiaz committed:
Use active and buffer hits in S1 calculation

gondiaz committed:
Add filter and warning to delayed hits

gondiaz committed:
Change S1 simulation procedure

S1 simulation was performed by sampling the s1-times distribution, but
if the S1 total pes is large the sampling might produce a memory
overflow. In this commit the simulation is changed from sampling to
computing the S1 distribution in the waveform by using the
distribution pdf.

gondiaz committed:
Update tests

MiryamMV commented on invisible_cities/detsim/detsim_waveforms.py:
```diff
@@ -1,37 +1,51 @@
+    Creates the s1-waveform based on the pdf at :s1_times_pdf:
+
+    Parameters:
+        :s1_pes_at_pmts: np.ndarray
```
This variable is called different in the definition of the function.

MiryamMV commented on invisible_cities/detsim/detsim_waveforms.py:
```diff
@@ -1,37 +1,51 @@
+            tmin + buffer_length
+        :bin_width:
+            waveform bin width
+        :tmin: float
```
Same here.

gondiaz committed:
Fix arg naming

MiryamMV reviewed (approved):
This PR raises a warning instead of filtering the delayed hits after `max_time`. It also computed the `s1` using the active and buffer volumes. Well done @gondiaz!

gonzaponte committed:
Add test file containing an event with an early low energy hit

This is a muon-induced xenon activation. The muon induces the recoil
of 137Xe which generates a very low energy hit. The 137Xe atom decays
much later producing the regular hits. The energy of the recoil hit
has been artificially reduced in order to have a more stable test.

The idea here is to have an event that "always" produces 0 electrons.

gonzaponte committed:
Add test to demonstrate that specific events cause an error

These events consist of a very low-energy hit (in this case produced
by the recoil of 137Xe) followed by very delayed regular hits. When
these hits are filtered out due to their high delay, only the
low-energy hit remains, which sometimes does not produce electrons,
leading to a crash.

gonzaponte committed:
Filter out dark events in detsim

Events that do not produce any photons are discarded

gonzaponte committed:
Extend the test to check more information

gonzaponte committed:
Check annotations in filter_hits_after_max_time
